sensors/simsensor_common/simsensor_reader.c

```diff
@@ -0,0 +1,284 @@
+				err = -1;
+				break;
+			}
+
+			clearerr(rd->scenarioFile);
+
+			/* Going to the beginning of the file and omitting file header */
+			if (fseek(rd->scenarioFile, rd->headerLen, SEEK_SET) != 0) {
+				err = -1;
+				break;
+			}
+			emptyIter--;
+
+			continue;
+		}
+
+		if (reader_getFieldLLong(rd->lineBuf, &nextField, &tmp) < 0) {
+			err = -1;
+			break;
+		}
+		sensorID = tmp;
+		actField = nextField;
+
+		/* Checking if we are parsing our header */
+		if ((sensorID & rd->sensorTypes) == 0) {
+			if (sensorID == READER_END_SCENARIO_INDICATOR) {
+				parsed.type = READER_STOP_EVENT;
+				if (eventQueue_enqueue(queue, &parsed) < 0) {
+					err = -1;
+				}
+
+				break;
+			}
+			emptyIter--;
+
+			continue;
+		}
+
+		if (reader_getFieldLLong(actField, &nextField, &tmp) < 0) {
+			err = -1;
+			break;
+		}
+		timestamp = tmp;
+		actField = nextField;
+
+		/* Increasing timeOffset on scenario timestamp decrease (as on file loop) */
+		if (timestamp < rd->timeLast && rd->timeLast != READER_TIMESTAMP_NOT_SET) {
+			rd->timeOffset += rd->timeLast - timestamp;
+		}
+		rd->timeLast = timestamp;
+		timestamp += rd->timeOffset;
+
+		if (timeStart == READER_TIMESTAMP_NOT_SET) {
+			timeStart = timestamp;
+		}
+
+		if (timestamp - timeStart > rd->timeHorizon) {
+			/* Go to start of the line */
+			if (fseek(rd->scenarioFile, -lineLen, SEEK_CUR) != 0) {
+				err = -1;
+			}
+			break;
+		}
+
+		switch (sensorID) {
+			case SENSOR_TYPE_ACCEL:
+				err = reader_accelDataParse(actField, timestamp, &parsed);
+				break;
+			case SENSOR_TYPE_BARO:
+				err = reader_baroDataParse(actField, timestamp, &parsed);
+				break;
+			default:
+				fprintf(stderr, "%s: Unknown sensor type: %d\n", __FUNCTION__, sensorID);
+				err = -1;
+				break;
+		}
+
+		if (err != 0) {
+			break;
+		}
+
+		if (eventQueue_enqueue(queue, &parsed) < 0) {
+			err = -1;
+			break;
+		}
+	}
+
+	return err;
+}
+
+
+void reader_close(simsens_reader_t *reader)
+{
+	if (reader == NULL) {
+		return;
+	}
+
+	if (reader->scenarioFile != NULL) {
+		fclose(reader->scenarioFile);
+		reader->scenarioFile = NULL;
+	}
+
+	free(reader->lineBuf);
+	reader->lineBuf = NULL;
+}
+
+
+int reader_open(simsens_reader_t *rd, const char *path, int sensorTypes, time_t timeHorizon)
+{
+	ssize_t lineLen;
+	time_t act_time;
+
+	if (rd == NULL || path == NULL) {
+		return -1;
+	}
+
+	rd->scenarioFile = fopen(path, "r");
+	if (rd->scenarioFile == NULL) {
+		return -1;
+	}
+
+	rd->lineBuf = NULL;
+	rd->bufLen = 0;
+	rd->sensorTypes = sensorTypes;
+	rd->timeHorizon = timeHorizon;
+	rd->timeLast = READER_TIMESTAMP_NOT_SET;
+	rd->headerLen = 0;
+
+	lineLen = getline(&rd->lineBuf, &rd->bufLen, rd->scenarioFile);
+	if (lineLen < 1) {
+		fclose(rd->scenarioFile);
+		return -1;
+	}
+
+	/* Checking if current line is header */
+	if (rd->lineBuf[0] == 's' || rd->lineBuf[0] == 'S') {
+		/* Read next line */
+		if (getline(&rd->lineBuf, &rd->bufLen, rd->scenarioFile) < 1) {
+			fclose(rd->scenarioFile);
+			return -1;
```

Comment:
Missing `free()`.
```suggestion
			free(rd->lineBuf);
			fclose(rd->scenarioFile);
			return -1;
```
If you leave `Free()` and `fclose()` without setting:
```c
	reader->scenarioFile = NULL;
	reader->lineBuf = NULL;
```
then `flose()` and `free()` can be called again in the `reader_close()` function.
I have an issue regarding `getline()`. In which repository to put it libphoenix or phoenix-rtos-project?

Reply:
Please report it in [phoenix-rtos-project issues](https://github.com/phoenix-rtos/phoenix-rtos-project/issues).